### ISSUE TITLE
mAP is too low

### ISSUE THREAD

**Issue by NguyenThang1501** [ultralytics/ultralytics#10226]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have trained a YOLOv8 model with a dataset that I labeled myself. The dataset consists of 1348 images and 50 labels related to various food items. After running for 100 epochs, the model yields very low precision, recall, and mAP. What should I do to improve the model?
![Screenshot 2024-04-20 162059](https://github.com/ultralytics/ultralytics/assets/109154036/82a2cbdc-e386-4b12-bde9-69b98dbcce08)

**Comment by glenn-jocher** [comment]
Hey there! 👋 Low mAP can stem from various factors, but don't worry, there are a few steps you can take to troubleshoot and hopefully see better results:

1. **Data Check**: Ensure your dataset is correctly labeled and there is a good balance across the various food item classes. Sometimes, imbalanced data can lead to poor model performance on less-represented classes.
2. **Augmentation**: Increase data diversity through augmentation (e.g., rotations, flips, color space adjustments). YOLOv8 directly supports several augmentation techniques which can be tweaked in your dataset YAML file.
3. **Hyperparameters**: Experiment with adjusting hyperparameters. For starting points, consider changing the learning rate or the number of epochs. Sometimes, longer training (more epochs) or using a learning rate scheduler can help.
4. **Model Architecture**: If your dataset is complex (and it sounds like it might be with 50 labels!), consider using a larger model variant of YOLOv8 which may capture details better than smaller models.

Here's a quick example of how to adjust augmentations in your YAML:

```yaml
# dataset.yaml
train: path/to/train/images
val: path/to/val/images

# Augmentations
mosaic: True
mixup: 0.2
hsv_h: 0.015  # hue HSV color space
hsv_s: 0.7   # saturation
hsv_v: 0.4   # value
```

And here's adjusting a learning rate schedule during training:

```bash
yolo train data=your_dataset.yaml model=yolov8s.pt epochs=300 lr0=0.01 lrf=0.2
```

Keep experimenting, and don't hesitate to reach out to the community for advice and share your progress! You've got this! 🚀

**Comment by glenn-jocher** [comment]
Hi there! 🌟

Absolutely, your model can learn to recognize a table with multiple dishes, even if most of your images feature only a single dish. The key is ensuring your dataset is diverse and annotated accurately, focusing on instances where multiple dishes are present on a table.

Keep in mind, model performance can lean towards the majority class in your dataset. However, you can strategically augment your data or artificially balance classes by giving more weight to the 'table with multiple dishes' during training. Here's a quick example to increase augmentation:

```yaml
# In your dataset.yaml
mosaic: True
mixup: 0.3
```

And to balance classes (if using custom training loop):

```python
# Adjust weights for your classes, higher for 'table with multiple dishes'
class_weights = torch.FloatTensor([1.0, 2.0]).to(device)  # Example values
```

This should help direct your model to better learn from the less represented but crucial class. Cheers to your model learning all about those delicious spreads of dishes! 🍽️

**Comment by NguyenThang1501** [comment]
Thank you for your enthusiasm. I would like to seek your opinion on the following matter: If I have a dish with multiple ingredients mixed together and I need model to detect the ingredients in it, should I assign the entire dish with multiple labels, or assign small details to each ingredient in the dish? Currently, I am only assigning that dish with one label as the main ingredient in the dish. But I am still hesitant in cases where the ingredients are equivalent in quantity. Can you suggest some solutions for this?

**Comment by glenn-jocher** [comment]
@NguyenThang1501 hey there! 🌟

Great question! If your goal is to identify individual ingredients in a dish, labeling each ingredient separately would be the ideal approach. This allows your model to learn to detect and distinguish between the various ingredients. For dishes where ingredients are mixed and roughly equal in quantity, you could still label each one. If visual differentiation is challenging, consider whether your project's scope includes training for such detailed recognition or if recognizing the dish as a whole suffices.

Here's a quick example of labeling for a dish with multiple ingredients:

- Image 1 (Pizza): Labels - "Cheese", "Pepperoni", "Bell Pepper", "Tomato Sauce"

If labeling each is too detailed for your current needs, it might be just as effective to label the dish based on its main characteristic or how it's commonly identified.

Whatever route you choose, consistency in labeling across your dataset is key! Happy modeling! 🚀
